File: backend/exercises/reach_bottle.py
import time
import math


class ReachBottleMetrics:
    def __init__(self, grasp_distance_threshold=80, movement_threshold=5):
        """
        grasp_distance_threshold: distance in pixels to consider hand "within grasp" of bottle
        movement_threshold: minimum movement (pixels) to detect start of movement
        """
        self.grasp_distance_threshold = grasp_distance_threshold
        self.movement_threshold = movement_threshold
        self.reset()

    def reset(self):
        self.start_signal_time = None
        self.movement_start_time = None
        self.reach_time = None
        self.reaction_time = None
        self.reach_completion_time = None

        self.movement_started = False
        self.reach_achieved = False

        self.initial_hand_pos = None
        self.last_hand_pos = None

        self.hand_positions = []  # List of (x,y) tuples
        self.direction_changes = 0
        self.prev_vector = None

    def update(self, current_time, hand_pos, bottle_pos):
        """
        Update metrics for the current frame.

        current_time: float, timestamp in seconds
        hand_pos: tuple (x,y) or None if hand not detected
        bottle_pos: tuple (x,y) or None if bottle not detected

        Returns dict with keys:
          - 'reaction_time': float or None
          - 'reach_time': float or None
          - 'trajectory_smoothness': int (direction changes count)
          - 'status': 'waiting' | 'waiting for detections' | 'waiting for movement' | 'moving toward bottle' | 'reached bottle'
        """
        if self.start_signal_time is None:
            return {'reaction_time': None, 'reach_time': None, 'trajectory_smoothness': 0, 'status': 'waiting'}

        if hand_pos is None or bottle_pos is None:
            return {'reaction_time': self.reaction_time, 'reach_time': self.reach_time,
                    'trajectory_smoothness': self.direction_changes,
                    'status': 'waiting for detections'}

        if self.initial_hand_pos is None:
            self.initial_hand_pos = hand_pos

        # Detect movement start
        dist_moved = math.dist(self.initial_hand_pos, hand_pos)
        if not self.movement_started and dist_moved > self.movement_threshold:
            self.movement_started = True
            self.movement_start_time = current_time
            self.reaction_time = self.movement_start_time - self.start_signal_time

        # Check reach
        if self.movement_started and not self.reach_achieved:
            dist_to_bottle = math.dist(hand_pos, bottle_pos)
            if dist_to_bottle <= self.grasp_distance_threshold:
                self.reach_achieved = True
                reach_duration = max(0.001, current_time - self.movement_start_time)
                self.reach_time = reach_duration
                self.reach_completion_time = current_time

        # Trajectory smoothness
        if self.last_hand_pos is not None:
            vec = (hand_pos[0] - self.last_hand_pos[0], hand_pos[1] - self.last_hand_pos[1])
            norm = math.hypot(vec[0], vec[1])
            if norm > 1e-6:
                vec_norm = (vec[0] / norm, vec[1] / norm)
                if self.prev_vector is not None:
                    dot = max(min(vec_norm[0] * self.prev_vector[0] + vec_norm[1] * self.prev_vector[1], 1.0), -1.0)
                    angle = math.acos(dot)
                    if angle > 0.785:  # ~45 degrees
                        self.direction_changes += 1
                self.prev_vector = vec_norm

        self.last_hand_pos = hand_pos
        self.hand_positions.append(hand_pos)

        if not self.movement_started:
            status = 'waiting for movement'
        elif not self.reach_achieved:
            status = 'moving toward bottle'
        else:
            status = 'reached bottle'

        return {
            'reaction_time': self.reaction_time,
            'reach_time': self.reach_time,
            'trajectory_smoothness': self.direction_changes,
            'status': status,
        }

    def start(self):
        self.reset()
        self.start_signal_time = time.time()

File: src/lib/exercises/reachBottle.ts
/**
 * Reach Bottle Exercise
 * Mirrors backend/exercises/reach_bottle.py ReachBottleMetrics
 */

export interface ReachBottleConfig {
  graspDistanceThreshold?: number; // pixels to consider hand "within grasp" of bottle
  movementThreshold?: number; // minimum movement (pixels) to detect start of movement
}

export interface ReachBottleState {
  startSignalTime: number | null;
  movementStartTime: number | null;
  reachTime: number | null;
  reactionTime: number | null;
  reachCompletionTime: number | null;
  movementStarted: boolean;
  reachAchieved: boolean;
  initialHandPos: { x: number; y: number } | null;
  lastHandPos: { x: number; y: number } | null;
  handPositions: Array<{ x: number; y: number }>;
  directionChanges: number;
  prevVector: { x: number; y: number } | null;
}

export interface ReachBottleResult {
  reactionTime: number | null;
  reachTime: number | null;
  trajectorySmoothness: number;
  status: 'waiting' | 'waiting for detections' | 'waiting for movement' | 'moving toward bottle' | 'reached bottle';
  passed: boolean;
}

export class ReachBottleMetrics {
  private config: Required<ReachBottleConfig>;
  private state: ReachBottleState;

  constructor(config: ReachBottleConfig = {}) {
    this.config = {
      graspDistanceThreshold: config.graspDistanceThreshold ?? 80,
      movementThreshold: config.movementThreshold ?? 5,
    };
    this.state = this.createInitialState();
  }

  private createInitialState(): ReachBottleState {
    return {
      startSignalTime: null,
      movementStartTime: null,
      reachTime: null,
      reactionTime: null,
      reachCompletionTime: null,
      movementStarted: false,
      reachAchieved: false,
      initialHandPos: null,
      lastHandPos: null,
      handPositions: [],
      directionChanges: 0,
      prevVector: null,
    };
  }

  reset(): void {
    this.state = this.createInitialState();
  }

  start(): void {
    this.reset();
    this.state.startSignalTime = Date.now() / 1000; // Convert to seconds
  }

  private distance(a: { x: number; y: number }, b: { x: number; y: number }): number {
    return Math.hypot(a.x - b.x, a.y - b.y);
  }

  update(
    currentTime: number, // in seconds
    handPos: { x: number; y: number } | null,
    bottlePos: { x: number; y: number } | null
  ): ReachBottleResult {
    if (this.state.startSignalTime === null) {
      return {
        reactionTime: null,
        reachTime: null,
        trajectorySmoothness: 0,
        status: 'waiting',
        passed: false,
      };
    }

    if (handPos === null || bottlePos === null) {
      return {
        reactionTime: this.state.reactionTime,
        reachTime: this.state.reachTime,
        trajectorySmoothness: this.state.directionChanges,
        status: 'waiting for detections',
        passed: false,
      };
    }

    if (this.state.initialHandPos === null) {
      this.state.initialHandPos = { ...handPos };
    }

    // Detect movement start
    const distMoved = this.distance(this.state.initialHandPos, handPos);
    if (!this.state.movementStarted && distMoved > this.config.movementThreshold) {
      this.state.movementStarted = true;
      this.state.movementStartTime = currentTime;
      if (this.state.startSignalTime !== null) {
        this.state.reactionTime = this.state.movementStartTime - this.state.startSignalTime;
      }
    }

    // Check reach
    if (this.state.movementStarted && !this.state.reachAchieved) {
      const distToBottle = this.distance(handPos, bottlePos);
      if (distToBottle <= this.config.graspDistanceThreshold) {
        this.state.reachAchieved = true;
        const reachDuration = Math.max(0.001, currentTime - (this.state.movementStartTime || currentTime));
        this.state.reachTime = reachDuration;
        this.state.reachCompletionTime = currentTime;
      }
    }

    // Trajectory smoothness
    if (this.state.lastHandPos !== null) {
      const vec = {
        x: handPos.x - this.state.lastHandPos.x,
        y: handPos.y - this.state.lastHandPos.y,
      };
      const norm = Math.hypot(vec.x, vec.y);
      if (norm > 1e-6) {
        const vecNorm = { x: vec.x / norm, y: vec.y / norm };
        if (this.state.prevVector !== null) {
          const dot = Math.max(-1, Math.min(1, vecNorm.x * this.state.prevVector.x + vecNorm.y * this.state.prevVector.y));
          const angle = Math.acos(dot);
          if (angle > 0.785) { // ~45 degrees
            this.state.directionChanges += 1;
          }
        }
        this.state.prevVector = vecNorm;
      }
    }

    this.state.lastHandPos = { ...handPos };
    this.state.handPositions.push({ ...handPos });

    let status: ReachBottleResult['status'];
    if (!this.state.movementStarted) {
      status = 'waiting for movement';
    } else if (!this.state.reachAchieved) {
      status = 'moving toward bottle';
    } else {
      status = 'reached bottle';
    }

    return {
      reactionTime: this.state.reactionTime,
      reachTime: this.state.reachTime,
      trajectorySmoothness: this.state.directionChanges,
      status,
      passed: this.state.reachAchieved,
    };
  }

  // Drawing helpers
  drawOverlay(
    ctx: CanvasRenderingContext2D,
    handPos: { x: number; y: number } | null,
    bottlePos: { x: number; y: number } | null,
    result: ReachBottleResult
  ): void {
    if (handPos === null || bottlePos === null) return;

    // Draw line from hand to bottle
    ctx.beginPath();
    ctx.moveTo(handPos.x, handPos.y);
    ctx.lineTo(bottlePos.x, bottlePos.y);
    ctx.strokeStyle = result.passed ? '#00FF00' : '#FF0000';
    ctx.lineWidth = 2;
    ctx.stroke();

    // Draw hand position
    ctx.beginPath();
    ctx.arc(handPos.x, handPos.y, 8, 0, 2 * Math.PI);
    ctx.fillStyle = result.passed ? '#00FF00' : '#FF6600';
    ctx.fill();

    // Draw bottle center
    ctx.beginPath();
    ctx.arc(bottlePos.x, bottlePos.y, 6, 0, 2 * Math.PI);
    ctx.fillStyle = '#00FF00';
    ctx.fill();

    // Draw status text
    ctx.fillStyle = '#FFFFFF';
    ctx.font = '14px sans-serif';
    ctx.fillText(`Status: ${result.status}`, 10, 30);
    if (result.reactionTime !== null) {
      ctx.fillText(`Reaction: ${(result.reactionTime * 1000).toFixed(0)}ms`, 10, 50);
    }
    if (result.reachTime !== null) {
      ctx.fillText(`Reach: ${(result.reachTime * 1000).toFixed(0)}ms`, 10, 70);
    }
    ctx.fillText(`Smoothness: ${result.trajectorySmoothness}`, 10, 90);
  }
}
